### PR TITLE
1/3 Unescape SMTP double dots, upgrade dependencies.

### PR DESCRIPTION
**Pull request by djanowski**
See https://github.com/nodemailer/nodemailer/issues/770.

@deitch If you want, after merging this I can work on cleaning up the whole codebase, then we can do a major release. What do you think?

### CODE REVIEW

**Comment by deitch** [comment]
> after merging this I can work on cleaning up the whole codebase, then we can do a major release. What do you think?

Who am I to argue? :-)

**Comment by deitch** [comment]
I am trying to wrap my head around the changes. Some I get, others not as much. And do we need to make changes to the README?

1. Tests: thank you for updating. Far too many PRs in projects I have been involved with that already have extensive tests... and the PRs neglect them.
2. `package.json`: Updating mailparser makes sense, and nodemailer really is only a dev dependency. So all good.
3. Rest in the code comments.

File: lib/index.js [hunk 1]
```diff
@@ -37,47 +37,50 @@ module.exports = {
 		var msgs = {}, msgid = 1, handlers = {}, processMsg, smtpServer, that;
 		smtpServer = smtp.createServer(smtpOptions);
 		smtpServer.on("startData",function(envelope){
-		  envelope.body = "";
+			envelope.body = "";
```
Comment (deitch) [inline lib/index.js]: Ironically, I have been switching all my JS to spaces over tabs. But I do not have any desire to be religious about it and aggressively change this!

File: lib/index.js [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 // get the authentication/sessionManager library and the authorization library
 var smtp = require('simplesmtp'), _ = require('underscore'), mimelib = require('mimelib-noiconv'), fs = require('fs'),
-MailParser = require("mailparser").MailParser,
+simpleParser = require("mailparser").simpleParser,
```
Comment (deitch) [inline lib/index.js]: New API to mailparser, ok.

File: lib/index.js [hunk 3]
```diff
@@ -37,47 +37,50 @@ module.exports = {
...
-				_.extend(email,mail_object || {});
-				// handle the http://tools.ietf.org/html/rfc2821#section-4.5.2 double-dot at the beginning of a line
-				// thanks to http://github.com/andris9 for pointing it out
-				//email.body = email.data.replace(/^\.\./mg,'.');
-				// is this email encoded quotedprintable? decode it
-				/*
-				if (email.headers && email.headers["Content-Transfer-Encoding"] && email.headers["Content-Transfer-Encoding"].toLowerCase() === "quoted-printable") {
-					email.body = mimelib.decodeQuotedPrintable(email.body);
+			// Handle the http://tools.ietf.org/html/rfc2821#section-4.5.2 double-dot at the beginning of a line
+			var unescapedBody = envelope.body.replace(/^\.\./mg, '.');
```
Comment (deitch) [inline lib/index.js]: Got it. Thanks for the ietf url in the comment.

File: lib/index.js [hunk 4]
```diff
@@ -37,47 +37,50 @@ module.exports = {
...
+				else {
+					_.extend(email,mail_object || {});
+					// handle the http://tools.ietf.org/html/rfc2821#section-4.5.2 double-dot at the beginning of a line
+					// email.body = email.text.replace(/^\.\./mg, '.');
+					// is this email encoded quotedprintable? decode it
+					/*
+					if (email.headers && email.headers["Content-Transfer-Encoding"] && email.headers["Content-Transfer-Encoding"].toLowerCase() === "quoted-printable") {
+						email.body = mimelib.decodeQuotedPrintable(email.body);
+					}
+					*/
```
Comment (deitch) [inline lib/index.js]: This stuff has been commented out for a very long time. Just delete it?

File: lib/index.js [hunk 5]
```diff
@@ -37,47 +37,50 @@ module.exports = {
...
+
+					// legacy support
+					email.headers.To = email.to.text;
+					email.headers.From = email.from.text;
+					email.body = email.text;
```
Comment (deitch) [inline lib/index.js]: What about the Xfolded and Xfoldedtab headers from the old lines 73-74?

File: lib/modules/logAll.js [hunk 6]
```diff
@@ -10,7 +10,7 @@ handler = function(addr,id,email) {
 	};
 	// structure the mail message
 	email.headers = casedown(email.headers);
-	var msg = ["From: "+email.headers.from, "To: "+email.headers.to, "Date: "+email.headers.date,"Subject: "+email.headers.subject, email.body];
+	var msg = ["From: "+email.from.text, "To: "+email.to.text, "Date: "+email.headers.date,"Subject: "+email.subject, email.text];
```
Comment (deitch) [inline lib/modules/logAll.js]: Is this part of the new mailparser API?

File: test/test.js [hunk 7]
```diff
@@ -18,19 +18,23 @@ function sendmail(to, subject, body, headers, cb) {
...
     subject: subject,
-    body:    body,
+    text:    body,
     headers: headers
   }, cb);
 }
 
 test('setup', function(t) {
-  smtpTransport = mailer.createTransport('SMTP',{
+  smtpTransport = mailer.createTransport({
```
Comment (deitch) [inline test/test.js]: Do we no longer specify transport in `mailer.createTransport`?

**Comment by deitch** [comment]
The only other question I have is the README. I guess since everything is encapsulated in the https://github.com/deitch/smtp-tester/blob/master/README.md#handlers as:

> Handlers that receive mail are passed three parameters.

> addr: Address to which the email was addressed, and for which the handler was bound. If this is a catch-all handler, then this is null.
> id: Internal ID of the email in this mail server process. Useful for removing messages or checking against something in our cache.
> email: JavaScript object of the email, containing "sender", "receivers", "data" (raw text), "headers" and "body".

Then as long as handlers still get that, and the `email` object contains `sender`, `receivers`, `data`, `headers` and `body`, then nothing has changed from a consumption perspective.

File: lib/index.js (hunk 1 above)
Comment (djanowski) [inline lib/index.js]: Let's switch to spaces in another PR.

File: lib/index.js (hunk 5 above)
Comment (djanowski) [inline lib/index.js]: I assumed those were hard-coded for tests? Didn't make sense to have them in here.

File: lib/index.js (hunk 4 above)
Comment (djanowski) [inline lib/index.js]: Done.

File: test/test.js (hunk 7 above)
Comment (djanowski) [inline test/test.js]: Apparently not per new Nodemailer.

**Comment by djanowski** [comment]
I updated Travis configuration to only test with LTS Boron (6.x), as that's the minimum version required by Nodemailer.

**Comment by djanowski** [comment]
> Then as long as handlers still get that, and the email object contains sender, receivers, data, headers and body, then nothing has changed from a consumption perspective.

OK, that's not what the tests were testing for, but we can make that work.

File: lib/index.js (hunk 5 above)
Comment (deitch) [inline lib/index.js]: Not really. I cannot recall the entire history here, but the issue is that headers are lowercased, which can create issues.

How did you remove them without tests failing?

File: test/test.js (hunk 7 above)
Comment (deitch) [inline test/test.js]: Yeah, it looks like you are right.

File: lib/index.js (hunk 5 above)
Comment (deitch) [inline lib/index.js]: Oh, I see. You rewrote that test as well.

File: test/test.js [hunk 8]
```diff
@@ -172,10 +176,10 @@ test('folded headers', function(t) {
...
   xfolded = 'This is a\r\n folded header', xfoldedtab = 'This is a\r\n\t tab folded header',
   handler = function(address, id, email) {
-    t.equal(email.headers.To,         recp);
-    t.equal(email.headers.From,       sender);
-    t.equal(email.headers.Xfolded,    xfolded.replace(/[\r\n\t]/g,''));
-    t.equal(email.headers.Xfoldedtab, xfoldedtab.replace(/[\r\n\t]/g,''));
+    t.equal(email.headers.To,                recp);
+    t.equal(email.headers.From,              sender);
+    t.equal(email.headers.get('xfolded'),    'This is a  folded header');
+    t.equal(email.headers.get('xfoldedtab'), 'This is a \t tab folded header');
```
Comment (deitch) [inline test/test.js]: Why the change in the test? Why not stick with `xfolded.replace(/[\r\n\t]/g,'')` and `xfoldedtab.replace(/[\r\n\t]/g,'')`?

These actually represent what it is supposed to do.

**Comment by deitch** [comment]
So, summary: all looks good except:

1. xfolded/tab tests. Why change it?
2. What did you mean by:

> that's not what the tests were testing for, but we can make that work

File: test/test.js (hunk 8 above)
Comment (djanowski) [inline test/test.js]: It's better to see the actual value expected as a literal instead of having to run some code in your head to know what the expected value is.

I changed this because this is how MailParser was behaving, but after your comment I went to read more about folding. If I understood correctly:

Newlines are generally not allowed in header values:

> A field body may be composed of printable US-ASCII characters
> as well as the space (SP, ASCII value 32) and horizontal tab (HTAB,
> ASCII value 9) characters (together known as the white space
> characters, WSP).  A field body MUST NOT include CR and LF except
> when used in "folding" and "unfolding", as described in section
> 2.2.3.

> Each header field is logically a single line of characters comprising
> the field name, the colon, and the field body.  For convenience
> however, and to deal with the 998/78 character limitations per line,
> the field body portion of a header field can be split into a
> multiple-line representation; this is called "folding".  The general
> rule is that wherever this specification allows for folding white
> space (not simply WSP characters), a CRLF may be inserted before any
> WSP.

So folding is a convenience for when you read email in your 80 char terminal. I tried it and in fact Nodemailer folds header values that are too long; MailParser unfolds them.

I think the correct test here would be to provide a really long value, and make sure it is parsed correctly into a single line.

**Comment by djanowski** [comment]
> What did you mean by:
> > that's not what the tests were testing for, but we can make that work

I updated the tests to reflect the spec of the email object.

**Comment by djanowski** [comment]
Also, I realized double-dot unescaping belongs to the SMTP server. Then I recalled that simplesmtp had been deprecated for ages, so upgraded ourselves to the newer smtp-server. And we don't have to do funny unescaping on our side (which would be wrong).

**Comment by deitch** [comment]
This is looking a lot better. Still a little confused, so going to try and summarize here, make sure I got it. And ask questions where needed.

1. Replace `simplesmtp` with `smtp-server`. OK
2. Cleans up handling messages in line with newer `smtp-server`. OK
3. Eliminate mailsplitter, since it is unused anyways. OK
4. No need for double-dot processing, since `smtp-server` does it for us. OK
5. Add tests for double-dots. OK
6. Removes the various xfolded headers in favour of a single one in the tests only, and eliminates need to process in run code (which is great). OK
7. Changes tests that expected upper-case headers for `To` and `From` to `to` and `from`.  Q
8. Changes tests that expected `email.subject` to email.headers.subject`. Q

So everything makes sense to me except the last 2. Actually, they do make sense to me, but does it change our API? Actually, I don't even mind. The API just says that the object should contain `"sender", "receivers", "data" (raw text), "headers" and "body".` so that is fine.

In that case, I think the only thing this is missing is changes to the README. 

1. It should change the headers in the example at https://github.com/deitch/smtp-tester#handlers to be all lower-case.
2. It should state explicitly that all headers are lower-cased.

Other than that, looks great.

**Comment by deitch** [comment]
And all tests pass too. Very nice!

**Comment by deitch** [comment]
Also, let's bump the version on this.

**Comment by djanowski** [comment]
> Changes tests that expected upper-case headers for `To` and `From` to `to` and `from`.

This upper-casing of headers was marked as "legacy support". Since we're going for a major bump, I thought we should just normalize to lowercase.

> Changes tests that expected `email.subject` to `email.headers.subject`.

`email.subject` wasn't in the spec, so I removed it per the same logic above. MailParser does copy some standard headers into the mail object itself (https://github.com/nodemailer/mailparser/blob/master/lib/mail-parser.js, search for "subject"). Should we follow?

> Also, let's bump the version on this.

I'd like to clean up the rest of the code in a separate PR before releasing a major, ok?

**Comment by deitch** [comment]
Looks good to me. We need to bump version though, if only to distribute to npm. Bump it, and let's merge in!

**Comment by djanowski** [comment]
Added #16 and #17 for easier reviewing, but if you are okay with everything, just merge #17.
